review/ark-blockstore-cstor/cstor.go

```diff
@@ -0,0 +1,526 @@
+        v1alpha1 "github.com/payes/maya/pkg/apis/openebs.io/v1alpha1"
+)
+
+const (
+	mayaAPIServiceName = "maya-apiserver-service"
+	backupCreatePath = "/latest/backups/"
+	restorePath = "/latest/restore/"
+	operator = "openebs"
+	casType = "cstor"
+	backupDir = "backups"
+)
+
+type cstorSnapPlugin struct {
+	Plugin cloudprovider.BlockStore
+        Log logrus.FieldLogger
+	K8sClient corev1.CoreV1Interface
+	config map[string]string
+	cl *cloudUtils
+	mayaAddr string
+	cstorServerAddr string
+        volumes   map[string]*Volume
+        snapshots map[string]*Snapshot
+}
+
+type Snapshot struct {
+        volID, backupName, namespace string
+}
+
+type Volume struct {
+        volname, casType, namespace, pvc, backupName string
+}
+
+func (p *cstorSnapPlugin) getServerAddress() string {
+        netInterfaceAddresses, err := net.InterfaceAddrs()
+
+        if err != nil {
+		p.Log.Errorf("Failed to get interface Address for ark server:%v", err)
+		return ""
+	}
+
+        for _, netInterfaceAddress := range netInterfaceAddresses {
+                networkIp, ok := netInterfaceAddress.(*net.IPNet)
+                if ok && !networkIp.IP.IsLoopback() && networkIp.IP.To4() != nil {
+                        ip := networkIp.IP.String()
+                        logrus.Infof("Resolved Host IP: " + ip)
+                        return ip + ":" + strconv.Itoa(RecieverPort)
+                }
+        }
+        return ""
+}
+
+func (p *cstorSnapPlugin) getMapiAddr() string {
+	sc, err := p.K8sClient.Services(operator).Get(mayaAPIServiceName, metav1.GetOptions{})
+        if err != nil {
+                p.Log.Errorf("Error getting IP Address for service - %s : %v", mayaAPIServiceName, err)
+		return ""
+        }
+
+        if len(sc.Spec.ClusterIP) != 0 {
+		return "http://" + sc.Spec.ClusterIP + ":" + strconv.FormatInt(int64(sc.Spec.Ports[0].Port), 10)
+        } else {
+		return ""
+	}
+}
+
+
+func (p *cstorSnapPlugin) Init(config map[string]string) error {
+        conf, err := rest.InClusterConfig()
+        if err != nil {
+		p.Log.Errorf("Failed to get cluster config", err)
+		return errors.New("Error fetching cluster config")
+        }
+
+	clientset, err := kubernetes.NewForConfig(conf)
+        if err != nil {
+		p.Log.Errorf("Error creating clientset", err)
+		return errors.New("Error creating k8s client")
+        }
+
+	p.K8sClient = clientset.CoreV1()
+	p.mayaAddr = p.getMapiAddr()
+	if p.mayaAddr == "" {
+		return errors.New("Error fetching OpenEBS rest client address")
+	}
+
+	p.cstorServerAddr = p.getServerAddress()
+	if p.cstorServerAddr == "" {
+		return errors.New("Error fetch cstorArkServer address")
+	}
+	p.config = config
+	if p.volumes == nil {
+		p.volumes = make(map[string]*Volume)
+	}
+	if p.snapshots == nil {
+		p.snapshots = make(map[string]*Snapshot)
+	}
+
+	p.cl = &cloudUtils {Log: p.Log}
+	return p.cl.InitCloudConn(config)
+}
+
+func (p *cstorSnapPlugin) GetVolumeID(pv runtime.Unstructured) (string, error) {
```

Comment:
This should be property of Snapshot structure defined in this file
We can move this to:
`pkg/ark/plugin/cstor/v1alpha1/snapshot.go`
```go
type snapshot struct {
  // ...
}

type snaphotBuilder struct {
  snapshot *snapshot
}

func SnapshotBuilder() *snapshotBuilder {
  return &snapshotBuilder{snapshot: &snapshot{}}
}

func (b *snapshotBuilder) WithVolumeID(volumeID string) *snapshotBuilder {
  b.snapshot.volumeID = volumeID
  return b
}

func (b *snapshotBuilder) WithNamespace(namespace string) *snapshotBuilder {
  b.snapshot.namespace = namespace
  return b
}

func (b *snapshotBuilder) WithUnstructured(u *unstructured.Unstructured) *snapshotBuilder {
  // ... refer to pkg/unstruct/v1alpha1 code in maya
  return b
}

func (b *snapshotBuilder) Build() *snapshot {
  return b.snapshot
}
```